internal/cli/clients/push.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	// returns staging ID
+	stage := &models.StageResponse{}
+	if err := json.Unmarshal(b, stage); err != nil {
+		return nil, err
+	}
+
+	return stage, nil
+}
+
+func (c *EpinioClient) logs(app models.AppRef, stageID string) (context.CancelFunc, error) {
+	c.ui.ProgressNote().V(1).Msg("Tailing application logs ...")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	selector := labels.NewSelector()
+
+	for _, req := range [][]string{
+		{"app.kubernetes.io/managed-by", "epinio"},
+		{models.EpinioStageIDLabel, stageID},
+		{"app.kubernetes.io/part-of", app.Org},
+		{"app.kubernetes.io/name", app.Name},
+	} {
+		req, err := labels.NewRequirement(req[0], selection.Equals, []string{req[1]})
+		if err != nil {
+			return cancelFunc, err
+		}
+		selector = selector.Add(*req)
+	}
+
+	err := tailer.Run(c.ui, ctx, &tailer.Config{
+		ContainerQuery:        regexp.MustCompile(".*"),
+		ExcludeContainerQuery: nil,
+		ContainerState:        "running",
+		Exclude:               nil,
+		Include:               nil,
+		Timestamps:            false,
+		Since:                 duration.LogHistory(),
+		AllNamespaces:         true,
+		LabelSelector:         selector,
+		TailLines:             nil,
+		Template:              tailer.DefaultSingleNamespaceTemplate(),
+		Namespace:             "",
+		PodQuery:              regexp.MustCompile(".*"),
+	}, c.KubeClient)
+	if err != nil {
+		return cancelFunc, errors.Wrap(err, "failed to start log tail")
+	}
+
+	return cancelFunc, nil
+}
+
+func (c *EpinioClient) waitForPipelineRun(app models.AppRef, id string) error {
+	c.ui.ProgressNote().KeeplineUnder(1).Msg("Running staging")
+
+	cs, err := tekton.NewForConfig(c.KubeClient.RestConfig)
+	if err != nil {
+		return err
+	}
+	client := cs.TektonV1beta1().PipelineRuns(deployments.TektonStagingNamespace)
+
+	if viper.GetInt("verbosity") > 0 {
+		s := c.ui.Progressf("Waiting for pipelinerun %s", id)
+		defer s.Stop()
+	}
+
+	return wait.PollImmediate(time.Second, duration.ToAppBuilt(),
+		func() (bool, error) {
+			l, err := client.List(context.TODO(), metav1.ListOptions{LabelSelector: models.EpinioStageIDLabel + "=" + id})
+			if err != nil {
+				return false, err
+			}
+			if len(l.Items) == 0 {
+				return false, nil
+			}
+			for _, pr := range l.Items {
+				// any failed conditions, throw an error so we can exit early
+				for _, c := range pr.Status.Conditions {
+					if c.IsFalse() {
+						return false, errors.New(c.Message)
+					}
+				}
+				// it worked
+				if pr.Status.CompletionTime != nil {
+					return true, nil
+				}
+			}
+			// pr exists, but still running
+			return false, nil
+		})
+}
+
+func (c *EpinioClient) waitForApp(app models.AppRef, id string) error {
+	c.ui.ProgressNote().KeeplineUnder(1).Msg("Creating application resources")
+	err := c.KubeClient.WaitUntilPodBySelectorExist(
+		c.ui, app.Org, fmt.Sprintf("app.kubernetes.io/name=%s,%s=%s", app.Name, models.EpinioStageIDLabel, id),
```

Comment:
Shouldn't this selector make use of this additional label? https://github.com/epinio/epinio/pull/399/files

Also the selector below (line 109) needs to filter the app pod. Now it's matching the staging pods too.

Otoh, the application should be considered "Running" when the deployment rollout is finished. I'm not sure what happens if the application has more than 1 pods. Does this code wait for all of them?

Reply:
I thought the selector in 109 wants to match the staging pod's logs? Not sure, I just moved the code to the new file.

I agree, waitForApp should wait for the deployment. The existing code waits for the pod to exist, then it waits for it to be running. (I think that could be one check.)
It checks all the pods that exist, so if the deployment wants ten pods and only 5 are there, it will check those 5.
Making sure the deployment is available would be the better check? https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#complete-deployment

Maybe, we do that in another story?

Reply:
yes let's do that in another story